R/tools.R

```diff
@@ -19,6 +19,84 @@ test_r <- function(...) {
   )
 }
 
+# From testthat https://github.com/r-lib/testthat/blob/v3.2.3/R/auto-test.R#L130
+starts_with <- function(string, prefix) {
+  substr(string, 1, nchar(prefix)) == prefix
+}
+
+check_if_includes_r_files <- function(path) {
+  r_files <- fs::dir_ls(path, recurse = TRUE, glob = "*.R$")
+
+  if (length(r_files) > 0) {
+    return(normalizePath(path))
+  }
+
+  NULL
+}
+
+# nolint start: line_length_linter
+#' Watch and automatically run R tests
+#'
+#' Watches R files in the `app` directory and `tests/testthat` directory for changes.
+#' When code files in `app` change, all tests are rerun. When test files change,
+#' only the changed test file is rerun.
+#'
+#' @param reporter `{testthat}` reporter to use.
+#'   If NULL, will use `testthat::default_reporter()` for tests when running all tests
+#'   and `testthat::default_compact_reporter()` for single file tests.
+#'   See [`{testthat}` reporters](https://testthat.r-lib.org/articles/reporters.htmlhttps://testthat.r-lib.org/reference/Reporter.html) for more details.
+#' @param hash Logical. Whether to use file hashing to detect changes. Default is TRUE.
+#'   If FALSE, file modification times are used instead.
+#'
+#' @return None. This function is called for side effects.
+#'
+#' @examples
+#' if (interactive()) {
+#'   # Watch files and automatically run tests when changes are detected
+#'   auto_test_r()
+#' }
+#' @export
+# nolint end
+auto_test_r <- function(reporter = NULL, hash = TRUE) {
```

Comment:
What if we allowed to pass additional arguments to `test_r`? I know we're mimicking the implementation of `testthat::auto_test`, but I think it could be more useful if we could also pass e.g. `filter` parameter.

Let's imagine our tests take 1 minute to finish running.

We're working on a feature that is quite well separated from the rest of the code we know what area should be impacted by the changes we're introducing and which test files are responsible for testing that area.

In that case it would make sense to allow using `filter` parameter to get feedback even faster, we can run the whole suite before committing our changes. 

```suggestion
auto_test_r <- function(reporter = NULL, hash = TRUE, ...) {
```
or
```suggestion
auto_test_r <- function(reporter = NULL, hash = TRUE, filter = NULL, <copy the rest of the args>) {
```

This is the same feedback I'd give if I was asked about `testthat::auto_test`. 😄 What do you think?

Reply:
The problem I can see here is that under the hood there are two functions (`test_dir` and `test_file`) that accept different sets of parameters. This inconsistency is on the level of `testthat`, so there is not much we can do about it. But I agree, that the `filter` param makes a lot of sense—the only challenge here is to filter the test files, so the tests are triggered only when the selected files are changed.

Reply:
I think it still makes sense if we don't filter source files:
- If you change source code:
  - if `filter = NULL` run all test files
  - if `filter != NULL` run selected files
- If you change test code:
  - if `filter = NULL` run changed test file 
  - if `filter = NULL` run changed test file

I think this would be the most consistent with testthat